ledger/src/blockstore.rs

```diff
@@ -3598,6 +3598,84 @@ impl Blockstore {
         self.get_slot_entries_in_block(slot, vec![(start_index, end_index)], slot_meta)
     }
 
+    /// Returns the merkle roots from the last `DATA_SHREDS_PER_FEC_BLOCK`
+    /// data shreds of a slot.
+    /// Will fail if:
+    ///     - Slot meta is missing
+    ///     - LAST_SHRED_IN_SLOT flag has not been received
+    ///     - There are missing shreds in the last fec set
+    /// If there are fewer than `DATA_SHREDS_PER_FEC_BLOCK` shreds in the slot
+    /// this will return all that are available.
+    fn get_last_merkle_roots(&self, slot: Slot) -> Result<Vec<Option<Hash>>> {
```

Comment:
I feel like this function is unnecessary. In particular allocating a `Vec` and buffering all Merkle roots is redundant.
We just need to check
* last_index >= 31
* last 32 shreds have the same Merkle roots. This does not need to allocate a vector. For example you can dedup the iterator:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.dedup
and verify it has only one element which is `Some(hash)`.